traits/trait_types.py

```diff
@@ -2283,16 +2283,40 @@ def __init__(self, *types, **metadata):
             if len(types) == 0:
                 types = [Trait(element) for element in default_value]
 
-        self.types = tuple([trait_from(type) for type in types])
+        self.types = tuple(trait_from(type) for type in types)
         self.init_fast_validate(ValidateTrait.tuple, self.types)
 
         if default_value is None:
-            default_value = tuple(
-                [type.default_value()[1] for type in self.types]
+            # Optimisation: if all child traits have a constant default value,
+            # we can use a constant default value too. Otherwise the default
+            # needs to be computed dynamically.
+            child_defaults = []
+            child_default_types = []
+            for child_trait in self.types:
+                default_type, default_value = child_trait.default_value()
+                child_default_types.append(default_type)
+                child_defaults.append(default_value)
+
+            constant_default = all(
+                dvt == DefaultValue.constant for dvt in child_default_types
             )
+            if constant_default:
+                self.default_value_type = DefaultValue.constant
+                default_value = tuple(child_defaults)
+            else:
+                self.default_value_type = DefaultValue.callable
+                default_value = self._get_default_value
 
         super().__init__(default_value, **metadata)
 
+    def _get_default_value(self, object):
+        # Dynamic default, used when at least one of the child traits requires
+        # a dynamic default.
+        return tuple(
+            inner_trait.default_value_for(object, "<inner_trait>")
```

Comment:
This part's annoying: we have to invent a name because `default_value_for` expects one, but the `DefaultValue.callable` default type doesn't pass any name along. So I chose something that should at least be non-ambiguous and can easily be searched for in the Traits source.

I couldn't find any way to actually get the name to show up in validation error messages. For example with `List` child traits, at the point where this dynamic default is used, it's then validated and at the validation step the `TraitList` objects are recreated with name matching the name of the actual trait; those recreated `TraitList` objects are what are actually used in the trait value.